drivers/include/adc_ng.h

```diff
@@ -0,0 +1,333 @@
+ *                          to achieve the requested resolution
+ *
+ * @retval  0               Success
+ * @retval  -ENOTSUP        No resolution equal to or greater than the requested
+ *                          found.
+ *
+ * Usage:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ * uint8_t res = 10, shift;
+ * int32_t sample;
+ *
+ * if (adc_ng_select_res(adc, &res, &shift) {
+ *     return -1;
+ * }
+ *
+ * if (adc_ng_init(adc, res, channel, &reference_voltage) {
+ *    return -1;
+ * }
+ *
+ * if (adc_ng_single(adc, &sample)) {
+ *     return -1;
+ * }
+ *
+ * adc_ng_off(adc);
+ *
+ * sample >>= shift;
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @note    In case of differential channels, the sample can be negative. In
+ *          that case an arithmetic right shift is required. Most (all?) C
+ *          compilers do so on signed types, but the C standard leaves it up
+ *          to the implementation to choose arithmetic or logical right shift.
+ *          For single ended ("normal") ADC channels samples are always positive
+ *          (or zero), and both flavours of right shift are fine.
+ */
+int adc_ng_select_res(uint8_t adc, uint8_t *res, uint8_t *shift);
+
+/** @} */
+
+/**
+ * @brief   Initialize and power up the ADC channel @p chan of device @p adc
+ *
+ * @param[in]       adc     ADC device to initialize a channel of
+ * @param[in]       chan    Channel to initialize
+ * @param[in]       res     Resolution to sample at
+ * @param[in,out]   ref     Reference voltage in mV to use (***see details!***)
+ *
+ * @retval  0               Success
+ * @retval  -ENOTSUP        Requested resolution not supported
+ * @retval  -ENXIO          No such channel
+ * @retval  -ERANGE         Requested reference voltage is higher than all
+ *                          available references
+ * @retval  -EALREADY       The ADC is already powered and configured
+ * @retval  <0              A driver specific error occurred
+ *
+ * @note    On success, the ADC is powered up and consumes additional power
+ *          until @ref adc_ng_off is called.
+ *
+ * The reference voltage to use is given with @p ref in millivolt. The driver
+ * will pick a reference voltage that is as close to @p ref as possible, but
+ * doesn't reduce the range. (The absolute value of the chosen reference is not
+ * smaller than the absolute value of the requested reference and it has the
+ * same sign as the requested reference.) The actually chosen reference voltage
+ * is stored in @p ref.
+ *
+ * E.g. if a device supports reference voltages -2.56V, -1.1V, 1.1V, and 2.56V
+ * and the user requests a reference voltage of 1.2V (`*ref == 1200`), the
+ * reference 2.56V is chosen even though 1.1V is a closer match. Otherwise
+ * input voltages above 1.1V would no longer be distinguishable. Similar if
+ * -1.3V is requested -2.56V is chosen over -1.1V.
```

Comment:
I really don't like this auto-magic.
Usually the hardware designer has thought of a specific supported reference voltage that is to be used with an ADC.
In the case of external voltage reference, there is no way the driver could know that value.
IMHO this should be part of `periph_conf.h`

Reply:
Could you provide more details on how this would look? Especially for the use case of two internal ADCs with a different set of reference voltages and one external with a third, distinct set of reference voltages.

Reply:
```
adc_common:
typedef struct adc_ref { 
int32_t voltage_mv; 
union{uint32 refconfv, void* refconfs} internal;
} adc_ref_t; 

MCU_adc:
const struct MCU_adc_dev {
 adc_ref_t references[NUMBER_OF_REFS];
} mcu_adc;
```
`int adc_ng_init(uint8_t adc, uint8_t chan, uint8_t res, const adc_ref_t  *ref)`

Initialisation will will be made with a pointer to the ref you want to use 


but i would actually go with

```
adc_init( *adc_dev, *adc_config)
```

the adc_config can be const and device specific and be prepared for the board / application.

and adc_mux could just get the same config.

Reply:
We can have just an "index" to identity a reference voltage. But we still need to add something to allow users to find what is the most suitable index. Sometimes looking it up for a specific ADC driver would be enough, but if the app is intended to compile and work on multiple boards whe need some abstraction.

Also, just exposing an array of the reference voltage values per ADC isn't sufficient. E.g. for the supply voltage as reference this may depend on the battary charge left ans the temperature, or the Seeduino UNI can run from 5V or 3.3V depending on a sliding button. The driver mostly can figure out the reference voltage at run time (e.g. by selecting the reference voltage of unkown value as reference and a stable internal reference voltage as input).

I don't mind having the fancy reference selection split out. But I really want to be able to write a "multimeter app" that works on any board with an ADC and a display supported by RIOT.

Reply:
But that's the thing, the reference voltage is usually not something you'd chose at run-time. ADC pins are usually hardwired to a certain functionality, so we can as well specify their voltage reference at the board level and don't have to make it part of the API.

Reply:
For the multimeter or the DSO use cases the run-time selection is crucial.

It can also be helpful for reading analog sensor. By swichting to a lower reference voltage, it is possible to pick up smaller changes. But going for a larger reference voltage, high sensor outputs won't oversample.